crates/adapters/src/format/mod.rs

```diff
@@ -399,24 +401,95 @@ impl dyn InputFormat {
     }
 }
 
-/// Parser that converts a raw byte stream into a stream of database records.
+/// A collection of records associated with an input handle.
 ///
-/// Note that the implementation can assume that either `input_fragment` or
-/// `input_chunk` will be called, but not both.
-pub trait Parser: Send {
+/// A [Parser] holds and adds records to an [InputBuffer].  The client, which is
+/// typically an [InputReader](crate::transport::InputReader), gathers one or
+/// more [InputBuffer]s and pushes them to the circuit when the controller
+/// requests it.
+pub trait InputBuffer: Send {
+    /// Pushes the `n` earliest buffered records into the circuit input
+    /// handle. If fewer than `n` are available, pushes all of them.  Discards
+    /// the records that are sent.  Returns the number sent.
+    fn flush(&mut self, n: usize) -> usize;
+
+    fn flush_all(&mut self) -> usize {
+        self.flush(usize::MAX)
+    }
+
+    /// Returns the number of buffered records.
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Removes all of the records from this input buffer and returns a new
+    /// [InputBuffer] that holds them. Returns `None` if this input buffer is
+    /// empty.
+    ///
+    /// This is useful for extracting the records from one of several parser
+    /// threads to send to a single common thread to be pushed later.
+    fn take(&mut self) -> Option<Box<dyn InputBuffer>>;
+}
+
+/// An implementation of [InputBuffer::flush] for a common pattern where worker
+/// threads append [InputBuffer]s to a common queue.
+pub(crate) fn flush_vecdeque_queue(
+    queue: &Mutex<VecDeque<Box<dyn InputBuffer>>>,
```

Comment:
Maybe worth experimenting with something lock free? (not in this PR)

Reply:
the lock free queues aren't necessarily faster or scale better and if they do usually just a little bit until they succumb to contention on some atomic

Reply:
yeah, I'm not a fan of them either. This lock might make it hard to scale the Kafka connector though. We'll probably need some solution eventually.

Reply:
we join the partition so we can guarantee some order on the input?